zebra-network/src/address_book.rs

```diff
@@ -463,6 +552,22 @@ impl AddressBook {
         self.by_addr.descending_values().cloned()
     }
 
+    /// Is this IP ready for a new outbound connection attempt?
+    /// Checks if the outbound connection with the most recent response at this IP has recently responded.
+    ///
+    /// Note: last_response times may remain live for a long time if the local clock is changed to an earlier time.
+    fn is_ready_for_connection_attempt_with_ip(
+        &self,
+        ip: &IpAddr,
+        chrono_now: chrono::DateTime<Utc>,
+    ) -> bool {
+        self.most_recent_by_ip
+            .as_ref()
+            .and_then(|most_recent_by_ip| most_recent_by_ip.get(ip))
+            .filter(|peer| peer.has_connection_recently_responded(chrono_now))
+            .is_none()
```

Comment:
I think the logic here might be inverted?

But I'm not sure, because the tests look like they are correct, and they are passing. Maybe we're missing a test for this specific case.

```rust
            .filter(|peer| !peer.has_connection_recently_responded(chrono_now))
```

Here's a refactor where that is easier to see:
```suggestion
        let Some(most_recent_by_ip) = self.most_recent_by_ip.as_ref() else {
            // if we're not checking IPs, any connection is allowed
            return true;
        }
        
        let Some(same_ip_peer) = most_recent_by_ip.get(ip) else {
            // If there's no entry for this IP, any connection is allowed
            return true;
        }
        
        !same_ip_peer.has_connection_recently_responded(chrono_now)
```

Reply:
> I think the logic here might be inverted?
> 
> But I'm not sure

The logic was correct, but your version is much clearer.